frontend/app/lib/server/s3.server.ts

```diff
@@ -0,0 +1,62 @@
+import {
+  GetObjectCommand,
+  PutObjectCommand,
+  S3Client
+} from '@aws-sdk/client-s3'
+import { streamToJson } from './utils.server'
+
+export class S3Service {
+  private static instance: S3Client | null = null
+  private client: S3Client
+  private bucketName: string
+  private fileKey: string
+
+  constructor(env: Env, walletAddress: string) {
+    // Use singleton pattern for S3 client
+    if (!S3Service.instance) {
+      S3Service.instance = new S3Client({
+        region: env.AWS_REGION,
+        credentials: {
+          accessKeyId: env.AWS_ACCESS_KEY_ID,
+          secretAccessKey: env.AWS_SECRET_ACCESS_KEY
+        }
+      })
+    }
+
+    this.client = S3Service.instance
+    this.bucketName = env.AWS_BUCKET_NAME
+    this.fileKey = `${decodeURIComponent(walletAddress)
+      .replace('$', '')
+      .replace('https://', '')}.json`
```

Comment:
I feel this should be a parameter to each method (the "key"), instead of an instance variable.
Like `get(key)`, and `put(key, data)`.